service/history/api/respondactivitytaskcompleted/api.go

```diff
@@ -97,13 +101,25 @@ func Invoke(
 			}
 
 			if !isRunning ||
-				ai.StartedEventId == common.EmptyEventID ||
+				(!isCalledCompleteActivityById && ai.StartedEventId == common.EmptyEventID) ||
 				(token.GetScheduledEventId() != common.EmptyEventID && token.Attempt != ai.Attempt) ||
 				(token.GetVersion() != common.EmptyVersion && token.Version != ai.Version) {
 				return nil, consts.ErrActivityTaskNotFound
 			}
 
-			if _, err := mutableState.AddActivityTaskCompletedEvent(scheduledEventID, ai.StartedEventId, request); err != nil {
+			// We fabricate a started event only when the activity is not started yet and
+			// we need to force complete an activity
+			if isCalledCompleteActivityById && ai.GetRetryLastFailure() != nil {
+				_, err := mutableState.AddActivityTaskStartedEvent(ai, scheduledEventID,
+					uuid.New().String(),
+					req.GetCompleteRequest().GetIdentity(),
+					mutableState.GetMostRecentWorkerVersionStamp())
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			if _, err = mutableState.AddActivityTaskCompletedEvent(scheduledEventID, ai.StartedEventId, request); err != nil {
```

Comment:
I don't think `ai` here is mutated after the add call above.
Please get `ai` from mutable state again.

Reply:
Yes. I'm not sure if we update `ai` in place, and it's better not to make that assumption.